test/irmin-pack/concurrent.ml

```diff
@@ -0,0 +1,187 @@
+  let rec aux i =
+    if i = n then Lwt.return_unit
+    else
+      S.Commit.of_hash r h >>= function
+      | None -> Alcotest.fail "no hash"
+      | Some commit ->
+          let msg = Printf.sprintf "RO find %d %d" n i in
+          let tree = S.Commit.tree commit in
+          S.Tree.find tree [ key i ] >>= fun x ->
+          Alcotest.(check (option string)) msg (Some (value i)) x;
+          aux (i + 1)
+  in
+  aux 0
+
+let add rw tree i =
+  Log.debug (fun l -> l "[%d] Adding round %d" parent_pid i);
+  S.Tree.add tree [ key i ] (value i) >>= fun tree ->
+  S.Commit.v rw ~parents:[] ~info:(info ()) tree >|= fun c ->
+  Log.debug (fun l -> l "Commit %a" S.Commit.pp_hash c);
+  (tree, c)
+
+let ignore_parent_pid (_ : int) = ()
+
+let worker w_pipe r_pipe =
+  Log.debug (fun l -> l "Worker started %d" (Unix.getpid ()));
+  ignore_parent_pid (read r_pipe);
+  Log.debug (fun l -> l "Signal received from parent to create store");
+  S.Repo.v (config ~readonly:true ~fresh:false root) >>= fun ro ->
+  let rec aux round_nb =
+    if round_nb = nb_batch_writes then Lwt.return_unit
+    else (
+      write w_pipe;
+      let h = read_hash r_pipe in
+      Log.debug (fun l ->
+          l "Signal received from parent to read data %a" (Irmin.Type.pp Hash.t)
+            h);
+      S.ro_sync ro;
+      test_find_present ro (round_nb + 1) h >>= fun () -> aux (round_nb + 1) )
+  in
+  aux 0 >>= fun () -> S.Repo.close ro
+
+let concurrent_reads () =
+  let read_main, write_worker = Unix.pipe ()
+  and read_worker, write_main = Unix.pipe () in
+  init ();
+  Printexc.record_backtrace true;
+  match Lwt_unix.fork () with
+  | 0 -> (
+      Log.debug (fun l -> l "[%d] Worker creates child" (Unix.getpid ()));
+      match Lwt_unix.fork () with
+      | 0 -> worker write_worker read_worker >|= fun () -> exit 0
+      | pid ->
+          worker write_worker read_worker >>= fun () ->
+          wait pid >|= fun () ->
+          Log.debug (fun l -> l "Child %d died" pid);
+          exit 0 )
+  | pid ->
+      S.Repo.v (config ~readonly:false ~fresh:true root) >>= fun rw ->
+      for _ = 0 to 1 do
+        write write_main
+      done;
+      let rec aux tree round_nb =
+        if round_nb = nb_batch_writes then Lwt.return_unit
+        else (
+          for _ = 0 to 1 do
+            let pid = read read_main in
+            Log.debug (fun l -> l "Ack from %d" pid)
+          done;
+          add rw tree round_nb >>= fun (tree, c) ->
+          let h = S.Commit.hash c in
+          for _ = 0 to 1 do
+            write_hash write_main h
+          done;
+          aux tree (round_nb + 1) )
+      in
+      aux S.Tree.empty 0 >>= fun () ->
+      wait pid >>= fun () ->
+      Unix.close read_main;
+      Unix.close write_worker;
+      Unix.close read_worker;
+      Unix.close write_main;
+      S.Repo.close rw
+
+let check_os () =
+  if Sys.os_type = "Win32" then Lwt.return_unit else concurrent_reads ()
```

Comment:
Perhaps a comment here to say that the reason we can't run this on Windows is the use of `fork`.